Overriding template property with non-template produces confusing error

This code ([Try link](https://try.metalama.net/#N4IghiBcIK4M4EsB2BzABAWQKYBcwBswBbMAOgDEAnYrAdwHtKBrUgQTgAcsBjHOAbgA6SYWABGcHNV5puhOHDQAhMHCyscUhGJg4saSGgAqATy7suvYcGFo7aANpGsRDoT0BdW-Y4wx+BG40cUlpHDRkcIAFSnouShwTNGA0FFx+NDUcDIBfYTyRJEisSiQCWXlFAEkkKXoAExheBHokDS0dPQNlVXVNSm1dLGtvO19-QLR6ADcSgfr9SLQYuJLE5NT0zK2CnJAAGhAOKBAQHKA)):

```c#
using Metalama.Framework.Aspects;

abstract class BaseAttribute : TypeAspect
{
    [Template]
    public abstract int Property { get; set; }
}

internal class IntroductionAttribute : BaseAttribute
{
    public override int Property { get; set; }
}
```

Produces on Try:

> The compile-time project could not be compiled. In most cases, this is due to a problem in your code and can be diagnosed using the other reported errors. If, however, you believe this is due to a bug in Metalama, please report the issue and include diagnostic information available in '...'.
> Aspect.cs(12,2): error CS0115: 'IntroductionAttribute.Property': no suitable method found to override

In the base property is `virtual` instead of `abstract`, the error is even more confusing:

> Aspect.cs(1,1): error CS1061: 'IntroductionAttribute' does not contain a definition for 'Property' and no accessible extension method 'Property' accepting a first argument of type 'IntroductionAttribute' could be found (are you missing a using directive or an assembly reference?)

--- a/Metalama.Framework/src/Metalama.Framework.Engine/Aspects/TemplateClassMemberBuilder.cs
+++ b/Metalama.Framework/src/Metalama.Framework.Engine/Aspects/TemplateClassMemberBuilder.cs
@@ -3,6 +3,7 @@
 // Refer to LICENSE.md in the repository root for complete details.
 
 using Metalama.Framework.Advising;
+using Metalama.Framework.Engine.CodeModel;
 using Metalama.Framework.Engine.CompileTime;
 using Metalama.Framework.Engine.Diagnostics;
 using Metalama.Framework.Engine.SerializableIds;
@@ -264,6 +265,32 @@ internal sealed class TemplateClassMemberBuilder : ITemplateClassMemberBuilder
 
                 // Add or replace the template.
                 membersBuilder[memberKey] = aspectClassMember;
+
+                // When a derived class overrides an abstract declarative advice member from the base class,
+                // the base class entry (which has IsAbstract=true) needs to be updated to reflect that
+                // a concrete implementation now exists. Without this, the introduction machinery will reject
+                // the abstract template when it tries to use the base entry for the introduction.
+                // The base entry is kept because it carries the [Introduce] attribute, but its TemplateInfo
+                // is updated to be non-abstract so the template validation succeeds. At runtime, virtual
+                // dispatch ensures the override's template body is executed.
+                if ( memberSymbol.IsOverride
+                     && templateInfo.AttributeType is TemplateAttributeType.DeclarativeAdvice or TemplateAttributeType.InterfaceMember )
+                {
+                    var overriddenMember = memberSymbol.GetOverriddenMember();
+
+                    if ( overriddenMember != null )
+                    {
+                        var overriddenKey = overriddenMember.GetDocumentationCommentId();
+
+                        if ( overriddenKey != null && overriddenKey != memberKey
+                                                   && membersBuilder.TryGetValue( overriddenKey, out var baseMember )
+                                                   && baseMember.TemplateInfo.IsAbstract )
+                        {
+                            // Use the override's template info (which is non-abstract) for the base entry.
+                            membersBuilder[overriddenKey] = baseMember with { TemplateInfo = templateInfo };
+                        }
+                    }
+                }
             }
             else
             {

--- a/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/Introductions/Properties/DeclarativeAbstract_InheritedTemplate.cs
+++ b/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/Introductions/Properties/DeclarativeAbstract_InheritedTemplate.cs
@@ -7,14 +7,14 @@ using Metalama.Framework.Aspects;
 
 namespace Metalama.Framework.Tests.AspectTests.Tests.Aspects.Introductions.Properties.DeclarativeAbstract_InheritedTemplate;
 
-// Tests that the [Template] attribute is inherited from a base aspect when overriding a property.
-// When a base aspect declares an abstract [Template] property and a derived aspect overrides it
-// without explicitly adding [Template], the override should still be recognized as a template.
+// Tests that the [Introduce] attribute is inherited from a base aspect when overriding a property.
+// When a base aspect declares an abstract [Introduce] property and a derived aspect overrides it
+// without explicitly adding [Introduce], the override should still be recognized as a declarative introduction.
 // See https://github.com/metalama/Metalama/issues/821
 
 public abstract class BaseAttribute : TypeAspect
 {
-    [Template]
+    [Introduce]
     public abstract int Property { get; set; }
 }
 

--- a/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/Introductions/Properties/DeclarativeAbstract_InheritedTemplate.cs
+++ b/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/Introductions/Properties/DeclarativeAbstract_InheritedTemplate.cs
@@ -1,0 +1,40 @@
+// Copyright (c) 2020-2025 SharpCrafters s.r.o. and contributors.
+// SharpCrafters s.r.o. licenses this file to you under either the MIT license or a proprietary license, depending on the repository from which it was obtained.
+// Refer to LICENSE.md in the repository root for complete details.
+
+using System;
+using Metalama.Framework.Aspects;
+
+namespace Metalama.Framework.Tests.AspectTests.Tests.Aspects.Introductions.Properties.DeclarativeAbstract_InheritedTemplate;
+
+// Tests that the [Template] attribute is inherited from a base aspect when overriding a property.
+// When a base aspect declares an abstract [Template] property and a derived aspect overrides it
+// without explicitly adding [Template], the override should still be recognized as a template.
+// See https://github.com/metalama/Metalama/issues/821
+
+public abstract class BaseAttribute : TypeAspect
+{
+    [Template]
+    public abstract int Property { get; set; }
+}
+
+public class IntroductionAttribute : BaseAttribute
+{
+    public override int Property
+    {
+        get
+        {
+            Console.WriteLine( "Introduced property getter." );
+
+            return 42;
+        }
+        set
+        {
+            Console.WriteLine( "Introduced property setter." );
+        }
+    }
+}
+
+// <target>
+[Introduction]
+internal class TargetClass { }

--- a/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/Introductions/Properties/DeclarativeAbstract_InheritedTemplate.t.cs
+++ b/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/Introductions/Properties/DeclarativeAbstract_InheritedTemplate.t.cs
@@ -1,0 +1,16 @@
+[Introduction]
+internal class TargetClass
+{
+  public global::System.Int32 Property
+  {
+    get
+    {
+      global::System.Console.WriteLine("Introduced property getter.");
+      return (global::System.Int32)42;
+    }
+    set
+    {
+      global::System.Console.WriteLine("Introduced property setter.");
+    }
+  }
+}

--- a/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/Introductions/Properties/DeclarativeVirtual_InheritedTemplate.cs
+++ b/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/Introductions/Properties/DeclarativeVirtual_InheritedTemplate.cs
@@ -1,0 +1,52 @@
+// Copyright (c) 2020-2025 SharpCrafters s.r.o. and contributors.
+// SharpCrafters s.r.o. licenses this file to you under either the MIT license or a proprietary license, depending on the repository from which it was obtained.
+// Refer to LICENSE.md in the repository root for complete details.
+
+using System;
+using Metalama.Framework.Aspects;
+
+namespace Metalama.Framework.Tests.AspectTests.Tests.Aspects.Introductions.Properties.DeclarativeVirtual_InheritedTemplate;
+
+// Tests that the [Template] attribute is inherited from a base aspect when overriding a virtual property.
+// When a base aspect declares a virtual [Template] property and a derived aspect overrides it
+// without explicitly adding [Template], the override should still be recognized as a template.
+// See https://github.com/metalama/Metalama/issues/821
+
+public abstract class BaseAttribute : TypeAspect
+{
+    [Template]
+    public virtual int Property
+    {
+        get
+        {
+            Console.WriteLine( "Base property getter (wrong)." );
+
+            return 0;
+        }
+        set
+        {
+            Console.WriteLine( "Base property setter (wrong)." );
+        }
+    }
+}
+
+public class IntroductionAttribute : BaseAttribute
+{
+    public override int Property
+    {
+        get
+        {
+            Console.WriteLine( "Overridden property getter." );
+
+            return 42;
+        }
+        set
+        {
+            Console.WriteLine( "Overridden property setter." );
+        }
+    }
+}
+
+// <target>
+[Introduction]
+internal class TargetClass { }

--- a/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/Introductions/Properties/DeclarativeVirtual_InheritedTemplate.cs
+++ b/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/Introductions/Properties/DeclarativeVirtual_InheritedTemplate.cs
@@ -7,14 +7,14 @@ using Metalama.Framework.Aspects;
 
 namespace Metalama.Framework.Tests.AspectTests.Tests.Aspects.Introductions.Properties.DeclarativeVirtual_InheritedTemplate;
 
-// Tests that the [Template] attribute is inherited from a base aspect when overriding a virtual property.
-// When a base aspect declares a virtual [Template] property and a derived aspect overrides it
-// without explicitly adding [Template], the override should still be recognized as a template.
+// Tests that the [Introduce] attribute is inherited from a base aspect when overriding a virtual property.
+// When a base aspect declares a virtual [Introduce] property and a derived aspect overrides it
+// without explicitly adding [Introduce], the override should still be recognized as a declarative introduction.
 // See https://github.com/metalama/Metalama/issues/821
 
 public abstract class BaseAttribute : TypeAspect
 {
-    [Template]
+    [Introduce]
     public virtual int Property
     {
         get

--- a/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/Introductions/Properties/DeclarativeVirtual_InheritedTemplate.t.cs
+++ b/Metalama.Framework/src/tests/Metalama.Framework.Tests.AspectTests/Tests/Aspects/Introductions/Properties/DeclarativeVirtual_InheritedTemplate.t.cs
@@ -1,0 +1,16 @@
+[Introduction]
+internal class TargetClass
+{
+  public global::System.Int32 Property
+  {
+    get
+    {
+      global::System.Console.WriteLine("Overridden property getter.");
+      return (global::System.Int32)42;
+    }
+    set
+    {
+      global::System.Console.WriteLine("Overridden property setter.");
+    }
+  }
+}

